### PR TITLE
feat: enhances e2e shannon relay test to a true E2E

### PR DESCRIPTION
**Pull request by commoddity**
This PR enhances the E2E test used to send a Shannon relay to use [Dockertest](https://github.com/ory/dockertest) to spin up the PATH service in an ephemeral Docker container and send a test relay through it.

This makes the test a true end-to-end test of the PATH service as it uses the same Docker image built from the root `Dockerfile` that would be used in production deployment.

Besides updating the [E2E test file itself](https://github.com/buildwithgrove/path/blob/true-e2e-test/e2e/shannon_relay_test.go), this PR makes the following modification, required to make the E2E work:

**Enhanced Health Check**
The `/healthz` endpoint in the `router` package now reports component ready states. It will return `503 Service Unavailable` until all required components have reported they are ready to serve relays; once ready it will return `200 OK`.

**_Changes_**
- Adds a `health.Check` interface which allows components to report their ready state:
```go
Check interface {
    Name() string  // Name returns the name of the component being checked.
    IsAlive() bool // IsAlive returns true if the component is healthy, otherwise false.
}
```
- Modifies the `relayer.Protocol` interface to include these methods.
```go
type Protocol interface {
	Endpoints(ServiceID) (map[AppAddr][]Endpoint, error)
	SendRelay(Request) (Response, error)
	// All components that report their ready status to /healthz must implement the health.Check interface.
	health.Check
}
```
- Modifies `morse.Protocol` and `shannon.Protocol` to report when they have required data to send relays:
```go
// Name satisfies the HealthCheck#Name interface function
func (p *Protocol) Name() string {
	return "pokt-shannon"
}

// IsAlive satisfies the HealthCheck#IsAlive interface function
func (p *Protocol) IsAlive() bool {
	p.appCacheMu.RLock()
	defer p.appCacheMu.RUnlock()
	p.sessionCacheMu.RLock()
	defer p.sessionCacheMu.RUnlock()

	return len(p.appCache) > 0 && len(p.sessionCache) > 0
}
```

**_This is also a useful feature for k8s deployments as it allows probes to only switch over to new pods on deployment when all components in the new pods have reported they are ready._**

### CODE REVIEW

File: e2e/shannon_relay_test.go [hunk 1]
```diff
@@ -4,102 +4,96 @@ package e2e
...
+	cleanupPathDocker(m, pool, resource)
+
+	// Exit with the test result
+	os.Exit(exitCode)
+}
 
-// TODO_INCOMPLETE: add an action to the CI for running E2E tests, which at the minimum
-// includes using the e2e tag and defining secrets to be used as environment variables, e.g. gateway/signer private key and address.
-//
 // TODO_IMPROVE: use gocuke (github.com/regen-network/gocuke) for defining and running E2E tests.
```
Comment (commoddity) [inline e2e/shannon_relay_test.go:44]: @adshmh @Olshansk is this still a `TODO`? I don't know enough about `gocuke` to say for sure what the advantage is but seems like it's probably not needed at this point, WDYT?

Comment (Olshansk) [reply e2e/shannon_relay_test.go:44]: Let's remove it for now. 

It'll come back once we move to natural language E2E tests: https://github.com/pokt-network/poktroll/blob/main/e2e/tests/0_settlement.feature

File: router/health_check.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+func (c *healthCheck) getComponentReadyStates() map[string]bool {
+	readyStates := make(map[string]bool)
+	for _, component := range c.components {
+		readyStates[component.Name()] = component.IsReady()
+	}
+	return readyStates
+}
+
+func getStatus(readyStates map[string]bool) healthCheckStatus {
```
Comment (Olshansk) [inline router/health_check.go:92]: Add godocs for stuff like this: `E.g. Return not ready if at least one of the components is not ready`

**Review by Olshansk — CHANGES_REQUESTED** [review]
Pausing at a partial review because I'm tired but also would appreciate if you take a second pass at `health_check.go`.

Not sure if I didn't dive deep enough, or if there's an opportunity for improvement.

File: cmd/main.go [hunk 3]
```diff
@@ -85,12 +85,20 @@ func main() {
 		Relayer:           relayer,
 	}
 
-	apiRouter := router.NewRouter(gateway, config.GetRouterConfig(), logger)
+	// Until all components are ready, the `/healthz` endpoint will return a 503 Service
```
[the remark this answers is not in the document]
Comment (commoddity) [reply cmd/main.go:88]: This is the intended behaviour. 503 indicates the service is still initializing.

File: router/router.go [hunk 4]
```diff
@@ -19,10 +17,11 @@ const (
...
 	router struct {
-		mux     *http.ServeMux
-		gateway gateway
-		config  config.RouterConfig
-		logger  polylog.Logger
+		mux         *http.ServeMux
+		gateway     gateway
+		healthCheck *healthCheck
+		config      config.RouterConfig
+		logger      polylog.Logger
```
[the remark this answers is not in the document]
Comment (commoddity) [reply router/router.go:24]: I do but I generally do it in the exact opposite order 😅

File: router/router.go [hunk 5]
```diff
@@ -32,27 +31,31 @@ type (
 /* --------------------------------- Init -------------------------------- */
 
 // NewRouter creates a new router instance
-func NewRouter(gateway gateway, config config.RouterConfig, logger polylog.Logger) *router {
+func NewRouter(gateway gateway, healthCheckComponents []HealthCheckComponent, config config.RouterConfig, logger polylog.Logger) *router {
```
[the remark this answers is not in the document]
Comment (commoddity) [reply router/router.go:34]: In a later PR I actually create a `RouterParams` struct to simplify it to one arg. I can bump that up to this PR since it seems relevant here.

File: relayer/morse/protocol.go [hunk 6]
```diff
@@ -59,6 +59,14 @@ type Protocol struct {
 	sessionCacheMu sync.RWMutex
 }
 
+func (p *Protocol) Name() string {
+	return "morse-protocol"
```
[the remark this answers is not in the document]
Comment (commoddity) [reply relayer/morse/protocol.go]: Done.

File: relayer/shannon/protocol.go [hunk 7]
```diff
@@ -60,6 +60,14 @@ type Protocol struct {
 	sessionCacheMu sync.RWMutex
 }
 
+func (p *Protocol) Name() string {
+	return "shannon-protocol"
```
[the remark this answers is not in the document]
Comment (commoddity) [reply relayer/shannon/protocol.go]: Done.

File: relayer/relayer.go [hunk 8]
```diff
@@ -81,6 +81,8 @@ type Response struct {
 // - Morse: in the relayer/morse package, and
 // - Shannon: in the relayer/shannon package.
 type Protocol interface {
+	Name() string
+	IsReady() bool
```
[the remark this answers is not in the document]
Comment (commoddity) [reply relayer/relayer.go]: Not really possible as we need to access the data inside the protocol cache to be able to know.

File: router/health_check.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	"os"
+
+	"github.com/pokt-network/poktroll/pkg/polylog"
+)
+
+type healthCheckStatus string
+
+const (
+	statusReady    healthCheckStatus = "ok"
+	statusNotReady healthCheckStatus = "initializing"
```
[the remark this answers is not in the document]
Comment (commoddity) [reply router/health_check.go]: Done.

File: router/health_check.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+// HealthCheckComponent is an interface that must be implemented
+// by components that need to report their health status
+type HealthCheckComponent interface {
+	Name() string
+	IsReady() bool
```
[the remark this answers is not in the document]
Comment (commoddity) [reply router/health_check.go]: Opted for the latter option. TY.

File: router/health_check.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+type (
```
[the remark this answers is not in the document]
Comment (commoddity) [reply router/health_check.go:37]: Have updated comments/naming. PTAL for explanation of everything.

File: router/health_check.go (hunk 2 above)
Comment (commoddity) [reply router/health_check.go:116]: Done.

File: router/router_mock_test.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
+// Code generated by MockGen. DO NOT EDIT.
+// Source: router.go
+//
+// Generated by this command:
+//
+//	mockgen -source=router.go -destination=router_mock_test.go -package=router
```
[the remark this answers is not in the document]
Comment (commoddity) [reply router/router_mock_test.go:6]: > Do you mind adding a single target in the makefile for all of the gomocks?
> 
> E.g. https://github.com/pokt-network/poktroll/blob/main/Makefile#L510

Sure but would you mind if we do this in a separate PR as this one already has quite a. bit going on. I will add a TODO to the Makefile.

File: router/router.go [hunk 13]
```diff
@@ -19,40 +17,58 @@ const (
...
+		config config.RouterConfig
+
+		logger polylog.Logger
 	}
 	gateway interface {
 		HandleHTTPServiceRequest(ctx context.Context, httpReq *http.Request, w http.ResponseWriter)
 	}
 )
 
+type RouterParams struct {
```
Comment (Olshansk) [inline router/router.go:33]: Challenging this. Is it really necessary?

I think passing all of it to the factory function is easier one by one.

File: e2e/shannon_relay_test.go [hunk 14]
```diff
@@ -4,102 +4,97 @@ package e2e
...
-		serviceID relayer.ServiceID
-		relayID   string
-		httpReq   *http.Request
+		name         string
+		reqMethod    string
+		reqPath      string
+		serviceID    string
+		serviceAlias string
+		relayID      string
+		body         string
```
Comment (Olshansk) [inline e2e/shannon_relay_test.go:52]:
```suggestion

		expectedBodyResponse         string
```

File: router/health_check.go [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+// HealthCheck is an interface that must be implemented
+// by components that need to report their health status
+type HealthCheck interface {
+	// Name returns the name of the component being checked.
+	Name() string
+	// IsAlive returns true if the component is healthy, otherwise false.
+	IsAlive() bool
+}
+
+// All components that implement the HealthCheck interface must be registered here.
```
Comment (Olshansk) [inline router/health_check.go:22]: Please move this into the corresponding files where the interface is being implemented.

It's a way to tell the developer: "You didn't implement the interface correctly".

The library (i.e. where the interface is defined) shouldn't have knowledge of who is implementing git.

File: router/health_check.go [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+var _ HealthCheck = &morse.Protocol{}
+var _ HealthCheck = &shannon.Protocol{}
+
+// One of `ready` or `not_ready`.
```
Comment (Olshansk) [inline router/health_check.go:26]: Make comments ever green. Developers usually forget to update comments if the implementation changes and the comment is not immediately visible to them.

```suggestion
// The status of the health check component.
```

File: router/health_check.go [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+type healthCheckStatus string
+
+const (
+	// StatusReady indicates that all PATH components are ready
```
Comment (Olshansk) [inline router/health_check.go:30]:
```suggestion
	// StatusReady indicates that the PATH component is ready
```

File: router/health_check.go [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+	statusReady healthCheckStatus = "ready"
+	// StatusNotReady indicates that one or more PATH components
+	// are still initializing, ie. warming up caches, etc.
```
Comment (Olshansk) [inline router/health_check.go:33]:
```suggestion
	// are still initializing (e.g. warming up caches, etc)
```

File: router/health_check.go [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+	statusNotReady healthCheckStatus = "not_ready"
+)
+
+type (
+	// The healthChecker struct is used to check the health of all PATH
+	// components that require startup time to warm up their caches, etc.
```
Comment (Olshansk) [inline router/health_check.go:39]:
```suggestion
	// The healthChecker struct is used to store all PATH components whose
	// health needs to be periodically checked.
```

File: relayer/shannon/protocol.go [hunk 20]
```diff
@@ -60,6 +60,19 @@ type Protocol struct {
 	sessionCacheMu sync.RWMutex
 }
 
+func (p *Protocol) Name() string {
```
Comment (Olshansk) [inline relayer/shannon/protocol.go:63]:
```suggestion
// Name satisfies the HealthCheck#Name interface function
func (p *Protocol) Name() string {
```

File: relayer/shannon/protocol.go [hunk 21]
```diff
@@ -60,6 +60,19 @@ type Protocol struct {
...
+	return "pokt-shannon"
+}
+
+func (p *Protocol) IsAlive() bool {
```
Comment (Olshansk) [inline relayer/shannon/protocol.go:67]:
```suggestion
// Name satisfies the HealthCheck#IsAlive interface function
func (p *Protocol) IsAlive() bool {
```

File: relayer/relayer.go [hunk 22]
```diff
@@ -83,6 +83,14 @@ type Response struct {
 type Protocol interface {
 	Endpoints(ServiceID) (map[AppAddr][]Endpoint, error)
 	SendRelay(Request) (Response, error)
+	ProtocolCache
+}
+
+// ProtocolCache reports if the protocol's cache is warm,
+// i.e. if it has the required data to serve requests.
+type ProtocolCache interface {
+	Name() string
```
Comment (Olshansk) [inline relayer/relayer.go:92]: 1. Why don't we just import/embed the HealthCheck interface in here (if that's the goal)
2. Alternatively, maybe this should be rename to `IsWarm` or `IsReady` instead of `IsAlive` (that's more of a health check thing)

File: e2e/docker_test.go [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+	containerExtraHost   = "host.docker.internal:host-gateway"
+	timeoutSeconds       = 120
+)
+
+var (
+	containerConfigMount     = filepath.Join(os.Getenv("PWD"), configFilePath) + configMountPoint
+	containerPortAndProtocol = internalPathPort + "/tcp"
+)
+
+func setupPathDocker() (*dockertest.Pool, *dockertest.Resource, string) {
```
Comment (Olshansk) [inline e2e/docker_test.go:37]: Please add godoc on how much this does (for future devs)

File: relayer/morse/protocol.go [hunk 24]
```diff
@@ -59,6 +59,19 @@ type Protocol struct {
 	sessionCacheMu sync.RWMutex
 }
 
+func (p *Protocol) Name() string {
```
Comment (Olshansk) [inline relayer/morse/protocol.go:62]: See my comments in `shannon/protocol`

File: router/health_check.go [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+		// Will default to `development` if not set in the image.
+		ImageTag string `json:"imageTag"`
+		// ReadyStates is a map of component names to their ready status
+		ReadyStates map[string]bool `json:"readyStates,omitempty"`
+	}
+)
+
+// healthCheckHandler returns the health status of PATH as a JSON response.
+//
+// It will return a 200 OK status code if all components are ready or
```
Comment (Olshansk) [inline router/health_check.go:60]: Nice

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: router/health_check.go (hunk 15 above)
Comment (commoddity) [reply router/health_check.go:22]: Not possible. It creates an import cycle between `router` and `shannon/morse`; that's why I put it where I did.

File: e2e/shannon_relay_test.go (hunk 14 above)
Comment (commoddity) [reply e2e/shannon_relay_test.go:52]: That's not what this variable is; it's the body that is sent as part of the request. The test doesn't care about the response currently; it only cares that at least one request succeeds.

File: router/health_check.go (hunk 17 above)
Comment (commoddity) [reply router/health_check.go:30]: I think indicating that it refers to multiple components being ready to serve relays is more accurate. Currently it is only the Protocol but in the future could be extended to many components that need to signal they are ready, eg. QoS.

File: router/health_check.go (hunk 16 above)
Comment (commoddity) [reply router/health_check.go:26]: Updated.

File: router/health_check.go (hunk 19 above)
Comment (commoddity) [reply router/health_check.go:39]: Updated.

File: relayer/shannon/protocol.go (hunk 20 above)
Comment (commoddity) [reply relayer/shannon/protocol.go:63]: Updated.

File: relayer/shannon/protocol.go (hunk 21 above)
Comment (commoddity) [reply relayer/shannon/protocol.go:67]: Updated.

File: relayer/relayer.go (hunk 22 above)
Comment (commoddity) [reply relayer/relayer.go:92]: Not possible; it leads to an import cycle between `router` and `relayer`. In addition, the `router` package should not be imported elsewhere; the interface should be defined in the package it's used in. 

In order to solve the import cycle issue I've lifted the health check to its own `health` package and imported elsewhere. This solves both this issue and [the issue of not being able to define the interface check in the `shannon/morse` packages.](https://github.com/buildwithgrove/path/pull/20#discussion_r1763920313)

File: router/health_check.go (hunk 15 above)
Comment (commoddity) [reply router/health_check.go:22]: [Solved this by moving the health check to its own package.](https://github.com/buildwithgrove/path/pull/20#discussion_r1764602485)

File: e2e/docker_test.go (hunk 23 above)
Comment (commoddity) [reply e2e/docker_test.go:37]: Done.

File: router/health_check.go (hunk 18 above)
Comment (commoddity) [reply router/health_check.go:33]: Done.

File: router/router.go (hunk 13 above)
Comment (commoddity) [reply router/router.go:33]: Sure. I've updated.

File: relayer/morse/protocol.go (hunk 24 above)
Comment (commoddity) [reply relayer/morse/protocol.go:67]: Done.

**Comment by commoddity** [comment]
@Olshansk In order to try and get the tests passing again I've updated the `shannon-sdk` version as fetching apps was not working. This also necessitated updating Go to `1.23`.

Note that they are still not passing, though for reasons that I am pretty sure are unrelated to PATH's code.

See my comment in the Discord channel. Let's try and debug this once you're online and once the E2E passes again we can merge this PR and get the proper E2E into the repo.

File: e2e/shannon_relay_test.go (hunk 14 above)
Comment (Olshansk) [reply e2e/shannon_relay_test.go:52]: #PUC

File: health/checker.go [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+	ImageTag string `json:"imageTag"`
+	// ReadyStates is a map of component names to their ready status
+	ReadyStates map[string]bool `json:"readyStates,omitempty"`
+}
+
+// healthCheckHandler returns the health status of PATH as a JSON response.
+//
+// It will return a 200 OK status code if all components are ready or
+// a 503 Service Unavailable status code if any component is not ready.
+
```
Comment (Olshansk) [inline health/checker.go:63]:
```suggestion
```

File: health/checker.go [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+const (
+	// StatusReady indicates that all PATH components are ready
+	statusReady healthCheckStatus = "ready"
+	// StatusNotReady indicates that one or more PATH components
+	// are still initializing (e.g. warming up caches, etc)
+	statusNotReady healthCheckStatus = "not_ready"
+)
+
+type (
+	// The health.Checker struct is used to store all PATH components whose
```
Comment (Olshansk) [inline health/checker.go:32]:
```suggestion
	// health.Checker struct is used to store all PATH components whose
```

File: health/checker.go [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+	// health needs to be checked to consider PATH ready to serve traffic.
+	Checker struct {
```
Comment (Olshansk) [inline health/checker.go:34]: Does this need to be publically exposed?

**Review by Olshansk — DISMISSED** [review]
Just a couple tiny comments / nits.

After we merge with main, the only outstanding item is ensuring the E2E passes (being discussed offline).

Will approve right after.

File: health/checker.go (hunk 28 above)
Comment (commoddity) [reply health/checker.go:34]: Yes, it's use in other packages (`main`, `relayer` and the relayer `morse` and `shannon` packages).

**Review by Olshansk — APPROVED** [review]
Approving because it's definitely in a good enough state while @commoddity is OOO!
